homeassistant/components/hlk_sw16.py

```diff
@@ -0,0 +1,230 @@
+        port = config[DOMAIN][device][CONF_PORT]
+
+        @callback
+        def reconnect():
+            """Schedule reconnect after connection has been lost."""
+            # Reset protocol binding before starting reconnect
+            SW16Device.set_hlk_sw16_protocol(None, device)
+
+            async_dispatcher_send(hass, SIGNAL_AVAILABILITY, False)
+
+            # If HA is not stopping, initiate new connection
+            if hass.state != CoreState.stopping:
+                _LOGGER.warning('disconnected from HLK-SW16, reconnecting')
+                hass.async_create_task(connect())
+
+        async def connect():
+            """Set up connection and hook it into HA for reconnect/shutdown."""
+            _LOGGER.info('Initiating HLK-SW16 connection')
+
+            try:
+                with async_timeout.timeout(CONNECTION_TIMEOUT,
+                                           loop=hass.loop):
+                    transport, protocol = await create_hlk_sw16_connection(
+                        host=host,
+                        port=port,
+                        disconnect_callback=reconnect,
+                        loop=hass.loop,
+                        logger=_LOGGER)
+                    _LOGGER.info(transport)
+
+            except (ConnectionRefusedError, TimeoutError, OSError,
+                    asyncio.TimeoutError) as exc:
+                reconnect_interval = config[DOMAIN][device]
+                [CONF_RECONNECT_INTERVAL]
+                _LOGGER.error(
+                    "Error connecting to HLK-SW16, reconnecting in %s",
+                    reconnect_interval)
+                # Connection lost make entities unavailable
+                async_dispatcher_send(hass, SIGNAL_AVAILABILITY, False)
+
+                hass.loop.call_later(reconnect_interval, reconnect, exc)
+                return
+
+            # Load platforms
+            for comp_name, comp_conf in switches.items():
+                hass.async_create_task(
+                    async_load_platform(hass, 'switch', DOMAIN,
+                                        (comp_name, comp_conf, device),
+                                        config))
+
+            # Bind protocol to command class to allow entities to send commands
+            SW16Device.set_hlk_sw16_protocol(protocol, device)
+
+            # handle shutdown of HLK-SW16 asyncio transport
+            hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP,
+                                       lambda x: transport.close())
+
+            _LOGGER.info('Connected to HLK-SW16')
+
+        hass.async_create_task(connect())
+
+    for device in config[DOMAIN]:
+        add_device(device)
+    return True
+
+
+class SW16Device(Entity):
+    """Representation of a HLK-SW16 device.
+
+    Contains the common logic for HLK-SW16 entities.
+    """
+
+    platform = None
+    _available = True
+    _protocol_reg = {}
+
+    def __init__(self, relay_name, device_port, device_id, name=None):
```

Comment:
Pass in the protocol when creating the entity, to make sure the entity has a protocol before being added to home assistant and registering protocol callbacks.